bundles/org.openhab.binding.solarman/src/main/java/org/openhab/binding/solarman/internal/updater/SolarmanChannelUpdater.java

```diff
@@ -0,0 +1,246 @@
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import javax.measure.Unit;
+import javax.measure.format.MeasurementParseException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.solarman.internal.defmodel.ParameterItem;
+import org.openhab.binding.solarman.internal.defmodel.Request;
+import org.openhab.binding.solarman.internal.defmodel.Validation;
+import org.openhab.binding.solarman.internal.modbus.SolarmanLoggerConnection;
+import org.openhab.binding.solarman.internal.modbus.SolarmanLoggerConnector;
+import org.openhab.binding.solarman.internal.modbus.SolarmanV5Protocol;
+import org.openhab.binding.solarman.internal.state.LoggerState;
+import org.openhab.binding.solarman.internal.typeprovider.ChannelUtils;
+import org.openhab.binding.solarman.internal.util.StreamUtils;
+import org.openhab.binding.solarman.internal.util.StringUtils;
+import org.openhab.core.library.types.DateTimeType;
+import org.openhab.core.library.types.DecimalType;
+import org.openhab.core.library.types.QuantityType;
+import org.openhab.core.library.types.StringType;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.types.State;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Catalin Sanda - Initial contribution
+ */
+@NonNullByDefault
+public class SolarmanChannelUpdater {
+    private final Logger logger = LoggerFactory.getLogger(SolarmanChannelUpdater.class);
+    private final StateUpdater stateUpdater;
+
+    public SolarmanChannelUpdater(StateUpdater stateUpdater) {
+        this.stateUpdater = stateUpdater;
+    }
+
+    public boolean fetchDataFromLogger(List<Request> requests, SolarmanLoggerConnector solarmanLoggerConnector,
+            SolarmanV5Protocol solarmanV5Protocol, Map<ParameterItem, ChannelUID> paramToChannelMapping,
+            LoggerState loggerState) {
+        try (SolarmanLoggerConnection solarmanLoggerConnection = solarmanLoggerConnector
+                .createConnection(!loggerState.isOffline())) {
+            logger.debug("Fetching data from logger");
+
+            Map<Integer, byte[]> readRegistersMap = requests.stream()
+                    .map(request -> solarmanV5Protocol.readRegisters(solarmanLoggerConnection,
+                            (byte) request.getMbFunctioncode().intValue(), request.getStart(), request.getEnd(),
+                            !loggerState.isOffline()))
+                    .reduce(new HashMap<>(), this::mergeMaps);
+
+            if (!readRegistersMap.isEmpty()) {
+                updateChannelsForReadRegisters(paramToChannelMapping, readRegistersMap);
+            }
+
+            return !readRegistersMap.isEmpty();
+        } catch (Exception e) {
```

Comment:
Can you catch the specific exceptions that can occur and route them to custom exceptions so the caller can decide on how to handle the situation.